### PR TITLE
feat(environment): Perform case insensitive comparison for config override names

### PR DESCRIPTION
**Pull request by lenny-intel**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Environment variable override names have to be exact case used in TOML file.

Issue Number:  #69

## What is the new behavior?

Environment variable override names can now be uppercase or an case

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

**Comment by iain-anderson** [comment]
OK with this. Because of the way override handling is implemented in device-sdk-c, when I update that it will only accept the original case or all-uppercase but I think that is fine.

**Comment by lenny-intel** [comment]
> OK with this. Because of the way override handling is implemented in device-sdk-c, when I update that it will only accept the original case or all-uppercase but I think that is fine.

I had same thought this weekend that either exact case or upper case only. I have made the tweak to handle this.

File: bootstrap/config/environment.go [hunk 1]
```diff
@@ -116,6 +127,68 @@ func (e *Environment) OverrideConfiguration(lc logger.LoggingClient, serviceConf
...
+			return lcKey, true
+		}
+	}
+
+	return "", false
+}
+
+func isAllUpperCase(key string) bool {
+	for _, ch := range key {
+		if unicode.IsLower(ch) && !unicode.IsUpper(ch) {
```
Comment (jim-wang-intel) [inline bootstrap/config/environment.go:184]: Do we need the extra `IsLower(ch)` comparison?  I thought `!Upper` is sufficient...

Comment (lenny-intel) [reply bootstrap/config/environment.go:184]: That is suppose to be IsLetter. Good catch!!

Comment (lenny-intel) [reply bootstrap/config/environment.go]: Fixed

**Comment by lenny-intel** [comment]
@jpwhitemn , no need to filter out special characters. If they are used, then it just doesn't match... I.e. non of our settings can have special characters. We use underscore because period is not a valid environment character, but the toml path uses period so we have to convert the underscore to a period.

File: bootstrap/config/environment.go [hunk 2]
```diff
@@ -116,6 +127,58 @@ func (e *Environment) OverrideConfiguration(lc logger.LoggingClient, serviceConf
...
+		ucMap[strings.ToUpper(path)] = path
+	}
+
+	return ucMap
+}
+
+// getKeyForMatchedPath searches for match of the environment variable name with the uppercase path (pathMap keys)
+// If matched found to original path (pathMap values) is returned as the "key"
+// For backward compatibility a case insensitive comparision is currently used.
+// TODO: For release v2.0.0 Change this to NOT check that `envVarName` is uppercase and only compare against uppercase
```
Comment (rsdmike) [inline bootstrap/config/environment.go:165]: Should we put this as an issue to address in the future?

Comment (lenny-intel) [reply bootstrap/config/environment.go:165]: Done https://github.com/edgexfoundry/go-mod-bootstrap/issues/71

**Pull request merged**